[BUGFIX] Improve handling of multidimensional turbine conditions

misi9170 opened the pull request:
#989, #785

misi9170 commented:
I'm now done with code changes (I think), but starting to work on the docs I think this is a good time to go ahead and improve turbine documentation across the board. I plan to add a Turbine "subfolder" to the documentation, and in it place (while reworking):
- [x] Turbine Operation Models
- [x] Floating turbine modeling #1125
- [x] Turbine library interface (add deprecation note, possibly warning in code) #901 
- [x] (new) multidimensional Cp/Ct definition information #1118 
- Turbine input file reference #1053, #1125

misi9170 commented:
After discussing with @dzalkind , I've decided that in this PR, I will include the bug fix for specifying multidimensional input conditions but not yet update reference values for the floating IEA 15MW turbine. In the mean time, I have updated various pieces of the documentation for floating turbines and indicated that the floating tilt table, along with the dependency of the thrust coefficient and power curves on wave height and wave period, should be treated as fictional. I will indicate this decision in related issues and discussions.

paulf81 commented on examples/examples_floating/001_floating_turbine_models.py:
```diff
@@ -89,7 +89,7 @@
     color="m",
     ls="--",
-    label="Floating (cp/ct not scaled by tilt)",
+    label="Floating (power/thrust coefficient not scaled by tilt)",
```
This could be a prior issue, but shouldn't the magenta dashed line be on top of the blue line in the 0th axis?

misi9170 replied:
I think this is currently correct. The idea here is that the black line represents the fixed-bottom NREL 5MW; the blue line represents what happens if you took the fixed NREL 5MW and put it on a floater (with a fictional level of added tilt for different wind speeds); and the magenta dashed line is the (fictional) case where the NREL 5MW power and thrust coefficient is defined as being on a floating platform, so no tilt correction to the power and thrust curve should be applied (as the effect of tilt is already inherently included).

paulf81 commented on examples/examples_multidim/003_multi_dimensional_cp_ct_TI.py:
```diff
@@ -0,0 +1,45 @@
+
+# Loop over different turbulence intensities using set()
+fig, axarr = plt.subplots(1, 3, sharex=True, figsize=(12, 4))
+for ti, col in zip([0.06, 0.10], ["k", "r"]):
```
The example works so I think is fine, but noting that TI is outside of the define bounds of the table, maybe comment if this is then fixed at the nearest or extrapolated?  Or this is covered in docs, just flagging

misi9170 replied:
Good point---the multidimensional turbine code always uses a nearest-neighbor approach. I've put a comment into the examples to this effect, and also added this information to the multidimensional turbine documentation.

paulf81 commented on examples/examples_multidim/003_multi_dimensional_cp_ct_TI.py:
```diff
@@ -0,0 +1,45 @@
+This example follows the previous example, but demonstrating how a multidimensional turbine can be
+used to model the effect of turbulence intensity on power and thrust coefficient.
+
+NOTE: The multi-dimensional power/thrust coefficient data used in this example is fictional for the
+purposes of facilitating this example.
```
Turbine 0's power is increased by turbulence, even near the knee of the power curve.  You point out a few times this is fake data but could also say "... and that is why the changes in power curve due to TI don't follow usual behaviors..."

paulf81 commented on docs/input_reference_main.md:
```diff
@@ -1,6 +1,6 @@
 # Main Input File Reference
 
-In additional to reinitializing {py:class}`FlorisInterface`, users can configure FLORIS
+In additional to calling the `set()` method on {py:class}`FlorisModel`, users can configure FLORIS
```
addition

paulf81 reviewed (approved):
I reviewed and ran the tests and went through and ran all examples.  I reviewed all the code and docs and all looks good to me.  Left a few minor comments

misi9170 commented:
@paulf81 Thanks for the review! I've now addressed your comments.

I also realized that we didn't have a way to store the multidimensional csv file in any user location---only in floris/turbine_library/, which was a bit unsatisfying. I've now made this possible by passing the `turbine_library_path` down to the `Turbine()` class, which instantiates the turbines and loads the multidimensional csv file, if it exists. This new approach is now used in examples/examples_multidim/003_multidimensional_cp_ct_TI.csv and also in the new test `test_set_multidim()` on floris_model_integration_test.py.

Assuming all examples run to completion, this PR will be ready for merge into the develop branch.

paulf81 commented:
agreed
